metadata-ingestion/src/datahub/ingestion/source/cassandra.py

```diff
@@ -0,0 +1,413 @@
+    )
+    username: str = Field(
+        default=None, description="The username credential."
+    )
+    password: str = Field(
+        default=None, description="The password credential."
+    )
+    excludeKeyspaces: List[str] = Field(
+        default=[], description="The keyspaces to exclude."
+    )
+
+
+# source reporter
+@dataclass
+class CassandraSourceReport(SourceReport):
+    filtered: List[str] = field(default_factory=list)
+
+    def report_dropped(self, index: str) -> None:
+        self.filtered.append(index)
+
+
+
+# -------------------------------------------------- the main source class -------------------------------------------------- #
+@platform_name("Cassandra")
+@config_class(CassandraSourceConfig)
+@support_status(SupportStatus.TESTING)
+@capability(SourceCapability.PLATFORM_INSTANCE, "Enabled by default")
+class CassandraSource(Source):
+
+    """
+    This plugin extracts the following:
+
+    - Metadata for tables
+    - Column types associated with each table column
+    - The keyspace each table belongs to
+    """
+
+    config: CassandraSourceConfig
+    report: CassandraSourceReport
+    cassandra_session: Session
+
+
+    def __init__(self, ctx: PipelineContext, config: CassandraSourceConfig):
+        self.ctx = ctx
+        self.config = config
+        self.report = CassandraSourceReport()
+
+        # attempt to connect to cass
+        ssl_context = SSLContext(PROTOCOL_TLSv1_2)
+        ssl_context.verify_mode = CERT_NONE
+        auth_provider = PlainTextAuthProvider(username=config.username, password=config.password)
+        cluster = Cluster([config.contact_point], port = config.port, auth_provider=auth_provider,ssl_context=ssl_context)
+        session = cluster.connect()
+        self.cassandra_session = session
+
+    @classmethod
+    def create(cls, config_dict, ctx):
+        config = CassandraSourceConfig.parse_obj(config_dict)
+        return cls(ctx, config)
+
+    def get_workunits_internal(
+            self,
+    ) -> Iterable[MetadataWorkUnit]:
+
+        # get all keyspaces and iterate through them
+        keyspaces = self.cassandra_session.execute(GET_KEYSPACES_QUERY)
+        keyspaces = sorted(keyspaces, key=lambda k: getattr(k, CASSANDRA_SYSTEM_SCHEMA_COLUMN_NAMES["keyspace_name"]))
+
+        for keyspace in keyspaces:
+            keyspace_name = getattr(keyspace, CASSANDRA_SYSTEM_SCHEMA_COLUMN_NAMES["keyspace_name"])
+            
+            # skip system keyspaces
+            if keyspace_name in SYSTEM_KEYSPACE_LIST:
+                continue
+            # skip keyspaces excluded by the config
+            if self.config.excludeKeyspaces and keyspace_name in self.config.excludeKeyspaces:
+                self.report.report_dropped(keyspace_name)
+                continue
+
+            # get all tables for this keyspace and emit their metadata
+            try:
+                yield from self._extract_tables_from_keyspace(keyspace_name)
+            except Exception as e:
```

Comment:
some classes don't have error handling, worth keeping though right?

Reply:
ie. does a hard failure help flag people's attention vs soft reporting?
if this is the case maybe if all keyspaces end up failing then i could throw an error for the caller to catch.